jax/api.py

```diff
@@ -1609,15 +1609,27 @@ def padded_spec(shape_spec):
       return tree_unflatten(out_tree, outs)
   return wrapped_fun
 
+NpDType = Any
 @curry
-def shapecheck(in_shapes, out_shape, fun: Callable):
+def shapecheck(in_shapes, out_shape, fun: Callable,
+               dtypes: Optional[Sequence[NpDType]] = None):
+  if dtypes is None:
+    dtypes = tree_map(lambda e: np.float32, in_shapes)
+  else:
+    pass
+    #assert len(dtypes) == len(in_shapes), (
+    #    f"Expected {len(in_shapes)} dtypes but got {len(dtypes)}.")
+
   _check_callable(fun)
   in_shapes, in_tree = tree_flatten(in_shapes)
+  in_dtypes, _ = tree_flatten(dtypes)
   in_shapes = map(masking.parse_spec, in_shapes)
   out_specs, out_spec_tree = tree_flatten(out_shape)
   out_specs = map(masking.parse_spec, out_specs)
   flat_fun, out_tree_thunk = flatten_fun_nokwargs(lu.wrap_init(fun), in_tree)
-  avals = map(partial(ShapedArray, dtype=np.float32), in_shapes)
+  avals = map(lambda typed_inshape: (
```

Comment:
`map` works with functions with multiple arguments, so you should be able to write the above as:
```
map(lambda in_shape, in_dtype: ShapedArray(in_shape, dtype=in_dtype), in_shapes, in_dtypes)
```